gems/quilt_rails/lib/quilt_rails/monkey_patches/active_support_reloader.rb

```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+require 'rails'
+
+# The default ActiveSupport::Reloader memoizes `@should_reload` for the
+# lifetime of each request. This is a problem for quilt_rails apps because:
+# - QuiltRails::ReactRenderable#render holds an exclusive lock until the Node
+#   server returns a result
+# - Any controller calls during Node rendering (e.g., GraphQL fetches) will
+#   hang if they try to obtain a lock
+# - Class unloading needs a lock, and so nested controller calls risk
+#   deadlocking whenever @should_reload is true
+#
+# Forcing `@should_reload` evaluation at the start of each thread prevents
+# nested controller calls from attempting class unloading.  This eliminates
+# one source of lock contention.
+#
+# The affected flow is:
+# - A developer saves a change to a Ruby file
+# - The developer refreshes their browser
+#
+# Rails processes the request:
+# - Thread0 - ActiveSupport::Reloader is called by middleware
+# - Thread0 - An exclusive class unloader lock is obtained, classes are
+#             unloaded, and the lock is released
+# - Thread0 - `Quilt::ReactRenderable#render_react` calls `reverse_proxy`,
+#             which grabs a general exclusive lock
+# - Node    - starts rendering, and calls out to a Rails controller for data
+# - Thread1 - ActiveSupport::Reloader is called by the second controller's
+#             middleware
+# - Thread1 - `@should_reload` is still true, so an attempt is made to grab an
+#              exclusive class unloader lock
+# - Thread1 - waits for Thread0 to release its lock; Thread0 never unlocks
+#             because it needs Thread1's result
+#
+module ActiveSupport
+  class Reloader
+    def self.check! # :nodoc:
+      @should_reload = check.call
```

Comment:
The upstream behaviour is to accomodate:

* A developer saves a change to a Ruby file
* Two requests are initiated
* Both requests enter executor
* First request decides it should reload, attempts to acquire lock, is stuck waiting for second to yield
* Second request hits `check.call`, gets false, proceeds without reloading
* First request completes reload and then proceeds

The fundamental problem is that the reload isn't known to be finished until the first request completes. You might be able to rearrange things inside the reloader so that concurrent requests are better able to overlap, while possibly having to do more re-work in the process.

Reply:
Thank you @matthewd for the review and suggestions. I guess even if we change the reloader we would have to wait for Rails 6.1 since I don't think it would be safe to backport this change, so I think we are stuck with this monkey patch for now.

I'm curious to know what this change would look like in Rails if we decide to do it.